<a href="https://colab.research.google.com/github/upcomingowner2023/ML/blob/main/FLP_2023.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import numpy as np
import pandas as pd


In [2]:
df=pd.read_csv("/content/supermarket.csv")

In [3]:
df.head()

Invoice ID Branch    City Customer type  Gender            Product line  \
0  750-67-8428      A   Delhi        Member  Female       Health and beauty   
1  226-31-3081      C  Mumbai        Normal  Female  Electronic accessories   
2  631-41-3108      A   Delhi        Normal    Male      Home and lifestyle   
3  123-19-1176      A   Delhi        Member    Male       Health and beauty   
4  373-73-7910      A   Delhi        Normal    Male       Sports and travel   

   Unit price  Quantity   Tax 5%     Total        Date  Time(in sec)  \
0       74.69         7  26.1415  548.9715  01-05-2019      0.547222   
1       15.28         5   3.8200   80.2200  03-08-2019      0.436806   
2       46.33         7  16.2155  340.5255  03-03-2019      0.557639   
3       58.22         8  23.2880  489.0480   1/27/2019      0.856250   
4       86.31         7  30.2085  634.3785  02-08-2019      0.442361   

       Payment  Cost Price  gross margin percentage  gross income  Rating  \
0      Ewallet      522.83                  26.1415       26.1415     9.1   
1         Cash       76.40                   3.8200        3.8200     9.6   
2  Credit card      324.31                  16.2155       16.2155     7.4   
3      Ewallet      465.76                  23.2880       23.2880     8.4   
4      Ewallet      604.17                  30.2085       30.2085     5.3   

   Paymen Type  
0          NaN  
1          NaN  
2          NaN  
3          NaN  
4          NaN

In [4]:
df.shape

(1000, 18)

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 18 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Invoice ID               1000 non-null   object 
 1   Branch                   1000 non-null   object 
 2   City                     1000 non-null   object 
 3   Customer type            1000 non-null   object 
 4   Gender                   1000 non-null   object 
 5   Product line             1000 non-null   object 
 6   Unit price               1000 non-null   float64
 7   Quantity                 1000 non-null   int64  
 8   Tax 5%                   1000 non-null   float64
 9   Total                    1000 non-null   float64
 10  Date                     1000 non-null   object 
 11  Time(in sec)             1000 non-null   float64
 12  Payment                  1000 non-null   object 
 13  Cost Price               1000 non-null   float64
 14  gross margin percentage  

In [6]:
df.describe()

Unit price     Quantity       Tax 5%        Total  Time(in sec)  \
count  1000.000000  1000.000000  1000.000000  1000.000000   1000.000000   
mean     55.672130     5.510000    15.379369   322.966749      0.642151   
std      26.494628     2.923431    11.708825   245.885335      0.133005   
min      10.080000     1.000000     0.508500    10.678500      0.416667   
25%      32.875000     3.000000     5.924875   124.422375      0.529861   
50%      55.230000     5.000000    12.088000   253.848000      0.638194   
75%      77.935000     8.000000    22.445250   471.350250      0.760417   
max      99.960000    10.000000    49.650000  1042.650000      0.874306   

       Cost Price  gross margin percentage  gross income      Rating  \
count  1000.00000              1000.000000   1000.000000  1000.00000   
mean    307.58738                15.379369     15.379369     6.97270   
std     234.17651                11.708825     11.708825     1.71858   
min      10.17000                 0.508500      0.508500     4.00000   
25%     118.49750                 5.924875      5.924875     5.50000   
50%     241.76000                12.088000     12.088000     7.00000   
75%     448.90500                22.445250     22.445250     8.50000   
max     993.00000                49.650000     49.650000    10.00000   

       Paymen Type  
count          0.0  
mean           NaN  
std            NaN  
min            NaN  
25%            NaN  
50%            NaN  
75%            NaN  
max            NaN

In [8]:
df.drop('Paymen Type', axis=1, inplace=True)
df.shape

(1000, 17)

In [10]:
df.columns

Index(['Invoice ID', 'Branch', 'City', 'Customer type', 'Gender',
       'Product line', 'Unit price', 'Quantity', 'Tax 5%', 'Total', 'Date',
       'Time(in sec)', 'Payment', 'Cost Price', 'gross margin percentage',
       'gross income', 'Rating'],
      dtype='object')

In [39]:
# Convert date column to datetime data type
df['Date'] = pd.to_datetime(df['Date'])

# Set date column as the index of the DataFrame
df.set_index('Date', inplace=True)

# Group the data by month and count the number of data points in each month
monthly_counts = df.resample('M').count()

monthly_counts['Total'].head()


Date
2019-01-31    352
2019-02-28    303
2019-03-31    345
Freq: M, Name: Total, dtype: int64

In [41]:
yearly_counts = df.resample('Y').count()
yearly_counts['Total'].head()

Date
2019-12-31    1000
Freq: A-DEC, Name: Total, dtype: int64

In [48]:
# Reset the index and create a new 'Date' column
df = df.reset_index()

In [49]:
# Group the data by month and aggregate the total sales for each month
df_monthly_sales = df.groupby(pd.Grouper(key='Date', freq='M'))['Total'].sum()
df_monthly_sales


Date
2019-01-31    116291.868
2019-02-28     97219.374
2019-03-31    109455.507
Freq: M, Name: Total, dtype: float64

By seeing the data we can conclude that , In 1st month the sales are high

In [58]:
# Group the data by year and aggregate the total sales for year
df_yearly_sales = df.groupby(pd.Grouper(key='Date', freq='Y'))['Total'].sum()
round(df_yearly_sales)

Date
2019-12-31    322967.0
Freq: A-DEC, Name: Total, dtype: float64

Total sales of the year is = **322967** rupees.

**Aggregating Functions:**

In [60]:
# Calculate the total sales for each product line
df_product_sales = df.groupby('Product line')['Total'].sum()
round(df_product_sales)

Product line
Electronic accessories    54338.0
Fashion accessories       54306.0
Food and beverages        56145.0
Health and beauty         49194.0
Home and lifestyle        53862.0
Sports and travel         55123.0
Name: Total, dtype: float64

We can conclude that most of the customers sales are done on Food and beverages.

In [61]:
# Calculate the average rating for each product line
df_product_rating = df.groupby('Product line')['Rating'].mean()
df_product_rating

Product line
Electronic accessories    6.924706
Fashion accessories       7.029213
Food and beverages        7.113218
Health and beauty         7.003289
Home and lifestyle        6.837500
Sports and travel         6.916265
Name: Rating, dtype: float64

Highest rating of the product is **"Food and beverages"**

**Manipulating Functions:**


In [68]:
# Add a new column with the cost of goods sold (COGS) for each record
df['COGS'] = df['Quantity'] * df['Cost Price']
df['COGS']

0      3659.81
1       382.00
2      2270.17
3      3726.08
4      4229.19
        ...   
995      40.35
996    9738.00
997      31.84
998      65.82
999    4328.66
Name: COGS, Length: 1000, dtype: float64

**Filtering (loc, iloc):**

In [70]:
# Select all records where the customer rating is above 4
df_high_rating = df.loc[df['Rating'] > 4]
df_high_rating

Date   Invoice ID Branch       City Customer type  Gender  \
0   2019-01-05  750-67-8428      A      Delhi        Member  Female   
1   2019-03-08  226-31-3081      C     Mumbai        Normal  Female   
2   2019-03-03  631-41-3108      A      Delhi        Normal    Male   
3   2019-01-27  123-19-1176      A      Delhi        Member    Male   
4   2019-02-08  373-73-7910      A      Delhi        Normal    Male   
..         ...          ...    ...        ...           ...     ...   
995 2019-01-29  233-67-5758      C     Mumbai        Normal    Male   
996 2019-03-02  303-96-2227      B  Bangalore        Normal  Female   
997 2019-02-09  727-02-1313      A      Delhi        Member    Male   
998 2019-02-22  347-56-2442      A      Delhi        Normal    Male   
999 2019-02-18  849-09-3807      A      Delhi        Member  Female   

               Product line  Unit price  Quantity   Tax 5%      Total  \
0         Health and beauty       74.69         7  26.1415  2195.8860   
1    Electronic accessories       15.28         5   3.8200    80.2200   
2        Home and lifestyle       46.33         7  16.2155   340.5255   
3         Health and beauty       58.22         8  23.2880  1956.1920   
4         Sports and travel       86.31         7  30.2085   634.3785   
..                      ...         ...       ...      ...        ...   
995       Health and beauty       40.35         1   2.0175    42.3675   
996      Home and lifestyle       97.38        10  48.6900  1022.4900   
997      Food and beverages       31.84         1   1.5920   133.7280   
998      Home and lifestyle       65.82         1   3.2910    69.1110   
999     Fashion accessories       88.34         7  30.9190  2597.1960   

     Time(in sec)      Payment  Cost Price  gross margin percentage  \
0        0.547222      Ewallet      522.83                  26.1415   
1        0.436806         Cash       76.40                   3.8200   
2        0.557639  Credit card      324.31                  16.2155   
3        0.856250      Ewallet      465.76                  23.2880   
4        0.442361      Ewallet      604.17                  30.2085   
..            ...          ...         ...                      ...   
995      0.573611      Ewallet       40.35                   2.0175   
996      0.719444      Ewallet      973.80                  48.6900   
997      0.556944         Cash       31.84                   1.5920   
998      0.647917         Cash       65.82                   3.2910   
999      0.561111         Cash      618.38                  30.9190   

     gross income  Rating     COGS  
0         26.1415     9.1  3659.81  
1          3.8200     9.6   382.00  
2         16.2155     7.4  2270.17  
3         23.2880     8.4  3726.08  
4         30.2085     5.3  4229.19  
..            ...     ...      ...  
995        2.0175     6.2    40.35  
996       48.6900     4.4  9738.00  
997        1.5920     7.7    31.84  
998        3.2910     4.1    65.82  
999       30.9190     6.6  4328.66  

[989 rows x 18 columns]

It means that 989 customers are 50% satisfied with thier products

In [72]:
# Select the first 10 records of the dataframe
df_first_10 = df.iloc[:10]
df_first_10

Date   Invoice ID Branch       City Customer type  Gender  \
0 2019-01-05  750-67-8428      A      Delhi        Member  Female   
1 2019-03-08  226-31-3081      C     Mumbai        Normal  Female   
2 2019-03-03  631-41-3108      A      Delhi        Normal    Male   
3 2019-01-27  123-19-1176      A      Delhi        Member    Male   
4 2019-02-08  373-73-7910      A      Delhi        Normal    Male   
5 2019-03-25  699-14-3026      C     Mumbai        Normal    Male   
6 2019-02-25  355-53-5943      A      Delhi        Member  Female   
7 2019-02-24  315-22-5665      C     Mumbai        Normal  Female   
8 2019-01-10  665-32-9167      A      Delhi        Member  Female   
9 2019-02-20  692-92-5582      B  Bangalore        Member  Female   

             Product line  Unit price  Quantity   Tax 5%      Total  \
0       Health and beauty       74.69         7  26.1415  2195.8860   
1  Electronic accessories       15.28         5   3.8200    80.2200   
2      Home and lifestyle       46.33         7  16.2155   340.5255   
3       Health and beauty       58.22         8  23.2880  1956.1920   
4       Sports and travel       86.31         7  30.2085   634.3785   
5  Electronic accessories       85.39         7  29.8865   627.6165   
6  Electronic accessories       68.84         6  20.6520  1734.7680   
7      Home and lifestyle       73.56        10  36.7800   772.3800   
8       Health and beauty       36.26         2   3.6260   304.5840   
9      Food and beverages       54.84         3   8.2260   690.9840   

   Time(in sec)      Payment  Cost Price  gross margin percentage  \
0      0.547222      Ewallet      522.83                  26.1415   
1      0.436806         Cash       76.40                   3.8200   
2      0.557639  Credit card      324.31                  16.2155   
3      0.856250      Ewallet      465.76                  23.2880   
4      0.442361      Ewallet      604.17                  30.2085   
5      0.770833      Ewallet      597.73                  29.8865   
6      0.608333      Ewallet      413.04                  20.6520   
7      0.484722      Ewallet      735.60                  36.7800   
8      0.718750  Credit card       72.52                   3.6260   
9      0.560417  Credit card      164.52                   8.2260   

   gross income  Rating     COGS  
0       26.1415     9.1  3659.81  
1        3.8200     9.6   382.00  
2       16.2155     7.4  2270.17  
3       23.2880     8.4  3726.08  
4       30.2085     5.3  4229.19  
5       29.8865     4.1  4184.11  
6       20.6520     5.8  2478.24  
7       36.7800     8.0  7356.00  
8        3.6260     7.2   145.04  
9        8.2260     5.9   493.56